DigitalLearningSolutions.Data/Services/DelegateUploadFileService.cs

```diff
@@ -0,0 +1,278 @@
+                delegateUserByAliasId.CandidateNumber != delegateUserByCandidateNumber.CandidateNumber)
+            {
+                delegateRow.Error = BulkUploadResult.ErrorReason.AliasIdInUse;
+                return;
+            }
+
+            var userToUpdate = delegateUserByCandidateNumber ?? delegateUserByAliasId;
+            if (userToUpdate == null)
+            {
+                if (!userService.IsEmailValidForCentre(delegateRow.Email!, centreId))
+                {
+                    delegateRow.Error = BulkUploadResult.ErrorReason.EmailAddressInUse;
+                    return;
+                }
+
+                RegisterDelegate(delegateRow, welcomeEmailDate, centreId);
+            }
+            else
+            {
+                ProcessPotentialUpdate(centreId, delegateRow, userToUpdate);
+            }
+        }
+
+        private DelegateUser? GetDelegateUserByCandidateNumberOrDefault(int centreId, string? candidateNumber)
+        {
+            return !string.IsNullOrEmpty(candidateNumber)
+                ? userDataService.GetDelegateUserByCandidateNumber(candidateNumber, centreId)
+                : null;
+        }
+
+        private DelegateUser? GetDelegateUserByAliasIdOrDefault(int centreId, string? aliasId)
+        {
+            return !string.IsNullOrEmpty(aliasId)
+                ? userDataService.GetDelegateUserByAliasId(aliasId, centreId)
+                : null;
+        }
+
+        private void ProcessPotentialUpdate(int centreId, DelegateTableRow delegateRow, DelegateUser delegateUser)
+        {
+            if (delegateRow.Email != delegateUser.EmailAddress &&
+                !userService.IsEmailValidForCentre(delegateRow.Email!, centreId))
+            {
+                delegateRow.Error = BulkUploadResult.ErrorReason.EmailAddressInUse;
+                return;
+            }
+
+            if (!RecordNeedsUpdating(delegateUser, delegateRow))
+            {
+                delegateRow.RowStatus = RowStatus.Skipped;
+                return;
+            }
+
+            UpdateDelegate(delegateRow, delegateUser);
+        }
+
+        private void UpdateDelegate(DelegateTableRow delegateRow, DelegateUser delegateUser)
+        {
+            userDataService.UpdateDelegate(
+                delegateUser.Id,
+                delegateRow.FirstName!,
+                delegateRow.LastName!,
+                delegateRow.JobGroupId!.Value,
+                delegateRow.Active!.Value,
+                delegateRow.Answer1,
+                delegateRow.Answer2,
+                delegateRow.Answer3,
+                delegateRow.Answer4,
+                delegateRow.Answer5,
+                delegateRow.Answer6,
+                delegateRow.AliasId,
+                delegateRow.Email!
+            );
+            delegateRow.RowStatus = RowStatus.Updated;
+        }
+
+        private static bool RecordNeedsUpdating(DelegateUser delegateUser, DelegateTableRow delegateRow)

```

Comment:
Given that the Validate() lives on DelegateTableRow, should this live there as well (e.g. bool MatchesRecord(DelegateUser user))?